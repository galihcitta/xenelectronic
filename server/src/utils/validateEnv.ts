import { port, cleanEnv } from 'envalid';

export const validateEnv = () => {
  cleanEnv(process.env, {
    PORT: port()
  });
};

// https://www.npmjs.com/package/envalid
