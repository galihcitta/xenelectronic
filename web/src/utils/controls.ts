import { Controls } from '../pages/Auth/types';

export const signInControls: Controls = {
  email: {
    label: 'Enter email',
    placeholder: '',
    type: 'email',
    id: 'signin-email',
    name: 'email',
    value: '',
    validationRules: {
      required: true,
      isEmail: true
    },
    valid: false,
    touched: false,
    errorMsg: ''
  },
  password: {
    label: 'Enter password',
    placeholder: '',
    type: 'password',
    id: 'signin-password',
    name: 'password',
    value: '',
    validationRules: {
      required: true
    },
    valid: false,
    touched: false,
    errorMsg: ''
  }
};

export const signUpControls: Controls = {
  username: {
    label: 'Enter username',
    placeholder: '',
    type: 'text',
    id: 'signup-username',
    name: 'username',
    value: '',
    validationRules: {
      required: true,
      minLength: 4,
      maxLength: 20
    },
    valid: false,
    touched: false,
    errorMsg: ''
  },
  email: {
    label: 'Enter email',
    placeholder: '',
    type: 'email',
    id: 'signup-email',
    name: 'email',
    value: '',
    validationRules: {
      required: true,
      isEmail: true
    },
    valid: false,
    touched: false,
    errorMsg: ''
  },
  password: {
    label: 'Enter password',
    placeholder: '',
    type: 'password',
    id: 'signup-password',
    name: 'password',
    value: '',
    validationRules: {
      required: true,
      isSecure: true,
      minLength: 8,
      maxLength: 20
    },
    valid: false,
    touched: false,
    errorMsg: ''
  }
};
