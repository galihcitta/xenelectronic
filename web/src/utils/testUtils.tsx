import React from 'react';
import { BrowserRouter, Router } from 'react-router-dom';
import { Provider } from 'react-redux';
import { store } from '../modules/store';
import { createMemoryHistory } from 'history';

const history = createMemoryHistory();

export const withRouter = <P extends object>(
  Component: React.ComponentType<P>
) => (props: P) => (
  <BrowserRouter>
    <Component {...props} />
  </BrowserRouter>
);

export const withHistory = <P extends object>(
  Component: React.ComponentType<P>
) => (props: P) => (
  <Router history={history}>
    <Component {...props} />
  </Router>
);

export const withProvider = <P extends object>(
  Component: React.ComponentType<P>
) => (props: P) => (
  <Provider store={store}>
    <BrowserRouter>
      <Component {...props} />
    </BrowserRouter>
  </Provider>
);
