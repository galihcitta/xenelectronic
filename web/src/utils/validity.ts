import { ValidationRules } from '../pages/Auth/types';

export const validate = (
  value: string,
  controlName: string,
  rules: ValidationRules
) => {
  let isValid = true;
  let errorMsg = '';

  if (rules.required) {
    isValid = value.trim() !== '' && isValid;
    if (!isValid && !errorMsg.length)
      errorMsg = `${controlName} field is required`;
  }

  if (rules.minLength) {
    isValid = value.length >= rules.minLength && isValid;
    if (!isValid && !errorMsg.length)
      errorMsg = `${controlName} must be at least ${rules.minLength} characters`;
  }

  if (rules.maxLength) {
    isValid = value.length <= rules.maxLength && isValid;
    if (!isValid && !errorMsg.length)
      errorMsg = `${controlName} must be no longer than ${rules.maxLength} characters`;
  }

  if (rules.isEmail) {
    const pattern = /[a-z0-9!#$%&'*+/=?^_`{|}~-]+(?:\.[a-z0-9!#$%&'*+/=?^_`{|}~-]+)*@(?:[a-z0-9](?:[a-z0-9-]*[a-z0-9])?\.)+[a-z0-9](?:[a-z0-9-]*[a-z0-9])?/;
    isValid = pattern.test(value) && isValid;
    if (!isValid && !errorMsg.length) errorMsg = `email is invalid`;
  }

  if (rules.isSecure) {
    // Minimum eight characters, at least one uppercase letter, one lowercase letter, one number and one special character
    const pattern = /^(?=.*[a-z])(?=.*[A-Z])(?=.*\d)(?=.*[@$!%*?&])[A-Za-z\d@$!%*?&]{8,}$/;
    isValid = pattern.test(value) && isValid;
    if (!isValid && !errorMsg.length) errorMsg = `password too weak`;
  }

  return {
    isValid,
    errorMsg: errorMsg.charAt(0).toUpperCase() + errorMsg.slice(1)
  };
};
