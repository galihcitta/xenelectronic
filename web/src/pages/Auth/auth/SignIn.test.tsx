import { render, fireEvent } from '@testing-library/react';
import { withRouter } from '../../../utils/testUtils';
import { _SignIn, SignInProps } from './SignIn.component';

const SignInWithRouter = withRouter(_SignIn);

describe('<SignIn', () => {
  const mockSignInStart = jest.fn();
  const mockClearErrors = jest.fn();

  const props: SignInProps = {
    signInStart: mockSignInStart,
    clearErrors: mockClearErrors,
    authError: null,
    isFetching: false
  };

  const setup = (props: SignInProps) => {
    const utils = render(SignInWithRouter(props));
    const form = utils.container.querySelector('form') as HTMLFormElement;
    const emailInput = utils.getByLabelText('Enter email') as HTMLInputElement;
    const passwordInput = utils.getByLabelText(
      'Enter password'
    ) as HTMLInputElement;
    const submitButton = utils.container.querySelector(
      'button[type="submit"]'
    ) as HTMLButtonElement;
    const error = utils.getByTestId('Error');

    return { ...utils, form, emailInput, passwordInput, submitButton, error };
  };

  it('renders correctly', () => {
    const { asFragment } = setup(props);
    expect(asFragment()).toMatchSnapshot();
  });

  describe('when `authError` prop is not null', () => {
    it('`Error` element contains correct style rules', () => {
      const { error, rerender } = setup(props);
      rerender(
        SignInWithRouter({ ...props, authError: 'Invalid credentials' })
      );

      expect(error).toHaveStyle('height: 40px');
      expect(error).toHaveStyle('opacity: 1');
    });

    it('`Error` element contains correct text', () => {
      const { getByText, rerender } = setup(props);
      rerender(
        SignInWithRouter({ ...props, authError: 'Invalid credentials' })
      );

      expect(getByText('Invalid credentials')).toBeInTheDocument();
    });
  });

  it('calls `SignInStart` action after submit', () => {
    const { emailInput, passwordInput, form } = setup(props);
    fireEvent.change(emailInput, { target: { value: 'johndoe@gmail.com' } });
    fireEvent.change(passwordInput, { target: { value: 'John123$' } });
    fireEvent.submit(form);

    expect(mockSignInStart).toHaveBeenCalledTimes(1);
  });

  it('contains a disabled button until credentials are valid', () => {
    const { emailInput, passwordInput, submitButton } = setup(props);
    fireEvent.change(emailInput, { target: { value: 'johndoe' } });
    fireEvent.change(passwordInput, { target: { value: 'a' } });

    expect(submitButton.disabled).toEqual(true);

    fireEvent.change(emailInput, { target: { value: 'johndoe@gmail.com' } });
    fireEvent.change(passwordInput, { target: { value: 'John123$' } });

    expect(submitButton.disabled).toEqual(false);
  });

  it('contains `Spinner` element when `isFetching` prop is true', () => {
    const { getByTestId } = setup({ ...props, isFetching: true });
    expect(getByTestId('Spinner')).toBeInTheDocument();
  });

  it('calls `clearErrors` action when component is unmounted', () => {
    const { unmount } = setup(props);
    unmount();

    expect(mockClearErrors).toHaveBeenCalled();
  });
});
