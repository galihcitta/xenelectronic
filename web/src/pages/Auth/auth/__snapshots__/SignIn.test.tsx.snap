// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SignIn renders correctly 1`] = `
<DocumentFragment>
  <form
    class="sc-bwzfXH fdSsPq"
  >
    <p
      class="sc-htpNat bclJaO"
    >
      Sign In
    </p>
    <span
      class="sc-EHOje cXVKmC"
      data-testid="Error"
    />
    <p
      class="sc-gzVnrw gATUFx"
    >
      <input
        class="sc-dnqmqq cDnSMn"
        id="signin-email"
        name="email"
        placeholder=""
        type="email"
        value=""
      />
      <span
        class="sc-iwsKbI ghpXNP"
        data-testid="ValidationIcon"
      >
        <i
          class="fas fa-check-circle"
        />
      </span>
      <label
        class="sc-htoDjs iaKsBG"
        for="signin-email"
      >
        Enter email
      </label>
      <span
        class="sc-gZMcBi gSNACN"
      />
    </p>
    <p
      class="sc-gzVnrw gATUFx"
    >
      <input
        class="sc-dnqmqq cDnSMn"
        id="signin-password"
        name="password"
        placeholder=""
        type="password"
        value=""
      />
      <span
        class="sc-iwsKbI ghpXNP"
        data-testid="ValidationIcon"
      >
        <i
          class="fas fa-check-circle"
        />
      </span>
      <label
        class="sc-htoDjs iaKsBG"
        for="signin-password"
      >
        Enter password
      </label>
      <span
        class="sc-gZMcBi gSNACN"
      />
    </p>
    <button
      class="sc-bZQynM eYyTpk"
      disabled=""
      type="submit"
    >
      Sign In
    </button>
    <p
      class="sc-bxivhb fgDZbI"
    >
      Don't have an account?
    </p>
    <a
      class="sc-ifAKCX jbVAZT"
      href="/account/register"
    >
      Register
    </a>
  </form>
</DocumentFragment>
`;
